// @croct/sdk/queue/capacityRestrictedQueue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CapacityRestrictedQueue = void 0;
var CapacityRestrictedQueue = /** @class */ (function () {
    function CapacityRestrictedQueue(queue, capacity) {
        this.queue = queue;
        this.capacity = capacity;
    }
    CapacityRestrictedQueue.prototype.all = function () {
        return this.queue.all();
    };
    CapacityRestrictedQueue.prototype.getCapacity = function () {
        return this.capacity;
    };
    CapacityRestrictedQueue.prototype.isEmpty = function () {
        return this.queue.isEmpty();
    };
    CapacityRestrictedQueue.prototype.length = function () {
        return Math.min(this.capacity, this.queue.length());
    };
    CapacityRestrictedQueue.prototype.peek = function () {
        return this.queue.peek();
    };
    CapacityRestrictedQueue.prototype.push = function (value) {
        if (this.queue.length() >= this.capacity) {
            throw new Error('Maximum queue capacity reached.');
        }
        this.queue.push(value);
    };
    CapacityRestrictedQueue.prototype.shift = function () {
        return this.queue.shift();
    };
    return CapacityRestrictedQueue;
}());
exports.CapacityRestrictedQueue = CapacityRestrictedQueue;
