// @croct/sdk/queue/inMemoryQueue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryQueue = void 0;
var tslib_1 = require("tslib");
var InMemoryQueue = /** @class */ (function () {
    function InMemoryQueue() {
        var _a;
        var values = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            values[_i] = arguments[_i];
        }
        this.queue = [];
        (_a = this.queue).unshift.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(values), false));
    }
    InMemoryQueue.prototype.all = function () {
        return this.queue.slice();
    };
    InMemoryQueue.prototype.getCapacity = function () {
        return Infinity;
    };
    InMemoryQueue.prototype.isEmpty = function () {
        return this.queue.length === 0;
    };
    InMemoryQueue.prototype.push = function (value) {
        this.queue.push(value);
    };
    InMemoryQueue.prototype.peek = function () {
        return this.queue[0] || null;
    };
    InMemoryQueue.prototype.shift = function () {
        var value = this.queue.shift();
        if (!value) {
            throw new Error('The queue is empty.');
        }
        return value;
    };
    InMemoryQueue.prototype.length = function () {
        return this.queue.length;
    };
    return InMemoryQueue;
}());
exports.InMemoryQueue = InMemoryQueue;
