// @croct/sdk/schema/contentSchemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.postDetails = void 0;
var validation_1 = require("../validation");
exports.postDetails = new validation_1.ObjectType({
    required: ['postId', 'title', 'publishTime'],
    properties: {
        postId: new validation_1.StringType({
            minLength: 1,
            maxLength: 100,
        }),
        url: new validation_1.StringType({
            format: 'url',
        }),
        title: new validation_1.StringType({
            minLength: 1,
            maxLength: 100,
        }),
        tags: new validation_1.ArrayType({
            items: new validation_1.StringType({
                minLength: 1,
                maxLength: 50,
            }),
            minItems: 1,
            maxItems: 10,
        }),
        categories: new validation_1.ArrayType({
            items: new validation_1.StringType({
                minLength: 1,
                maxLength: 50,
            }),
            minItems: 1,
            maxItems: 10,
        }),
        authors: new validation_1.ArrayType({
            items: new validation_1.StringType({
                minLength: 1,
                maxLength: 50,
            }),
            minItems: 1,
            maxItems: 10,
        }),
        publishTime: new validation_1.NumberType(),
        updateTime: new validation_1.NumberType(),
    },
});
