// @croct/sdk/schema/sdkFacadeSchemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sdkFacadeConfigurationSchema = void 0;
var validation_1 = require("../validation");
var contextSchemas_1 = require("./contextSchemas");
var sdkSchemas_1 = require("./sdkSchemas");
var loggerSchema_1 = require("./loggerSchema");
exports.sdkFacadeConfigurationSchema = new validation_1.ObjectType({
    required: ['appId'],
    properties: {
        appId: new validation_1.StringType({
            format: 'uuid',
        }),
        cid: new validation_1.StringType({
            pattern: /^[0-9a-f]{32}$/i,
        }),
        tokenScope: contextSchemas_1.tokenScopeSchema,
        debug: new validation_1.BooleanType(),
        test: new validation_1.BooleanType(),
        track: new validation_1.BooleanType(),
        logger: loggerSchema_1.loggerSchema,
        urlSanitizer: new validation_1.FunctionType(),
        eventMetadata: sdkSchemas_1.eventMetadataSchema,
        userId: new validation_1.StringType({
            minLength: 1,
        }),
        token: new validation_1.UnionType(new validation_1.StringType({
            pattern: /^[A-Za-z0-9-_=]+\.[A-Za-z0-9-_=]+\.?[A-Za-z0-9-_.+/=]*$/,
        }), new validation_1.NullType()),
        trackerEndpointUrl: new validation_1.StringType({
            format: 'url',
        }),
        evaluationEndpointUrl: new validation_1.StringType({
            format: 'url',
        }),
        bootstrapEndpointUrl: new validation_1.StringType({
            format: 'url',
        }),
    },
});
