// @croct/sdk/schema/sdkSchemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sdkConfigurationSchema = exports.eventMetadataSchema = void 0;
var validation_1 = require("../validation");
var contextSchemas_1 = require("./contextSchemas");
var loggerSchema_1 = require("./loggerSchema");
exports.eventMetadataSchema = new validation_1.ObjectType({
    maxProperties: 5,
    propertyNames: new validation_1.StringType({
        minLength: 1,
        maxLength: 20,
        format: 'identifier',
    }),
    additionalProperties: new validation_1.StringType({
        maxLength: 300,
    }),
});
exports.sdkConfigurationSchema = new validation_1.ObjectType({
    required: ['appId'],
    properties: {
        appId: new validation_1.StringType({
            format: 'uuid',
        }),
        cid: new validation_1.StringType({
            pattern: /^[0-9a-f]{32}$/i,
        }),
        tokenScope: contextSchemas_1.tokenScopeSchema,
        trackerEndpointUrl: new validation_1.StringType({
            format: 'url',
        }),
        evaluationEndpointUrl: new validation_1.StringType({
            format: 'url',
        }),
        bootstrapEndpointUrl: new validation_1.StringType({
            format: 'url',
        }),
        beaconQueueSize: new validation_1.NumberType({
            minimum: 0,
            integer: true,
        }),
        debug: new validation_1.BooleanType(),
        test: new validation_1.BooleanType(),
        logger: loggerSchema_1.loggerSchema,
        urlSanitizer: new validation_1.FunctionType(),
        eventMetadata: exports.eventMetadataSchema,
    },
});
