// @croct/sdk/schema/tokenSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tokenSchema = void 0;
var validation_1 = require("../validation");
exports.tokenSchema = new validation_1.ObjectType({
    required: ['headers', 'claims'],
    properties: {
        headers: new validation_1.ObjectType({
            required: ['typ', 'alg', 'appId'],
            properties: {
                typ: new validation_1.StringType(),
                alg: new validation_1.StringType(),
                kid: new validation_1.StringType(),
                appId: new validation_1.StringType({
                    format: 'uuid',
                }),
            },
        }),
        claims: new validation_1.ObjectType({
            required: ['iss', 'aud', 'iat'],
            properties: {
                iss: new validation_1.StringType(),
                aud: new validation_1.UnionType(new validation_1.StringType(), new validation_1.ArrayType({ items: new validation_1.StringType() })),
                iat: new validation_1.NumberType({
                    minimum: 0,
                }),
                sub: new validation_1.StringType({
                    minLength: 1,
                }),
                exp: new validation_1.NumberType({
                    minimum: 0,
                }),
                jti: new validation_1.StringType({
                    format: 'uuid',
                }),
            },
        }),
        signature: new validation_1.StringType(),
    },
});
