// @croct/sdk/channel/sandboxChannel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SandboxChannel = void 0;
var SandboxChannel = /** @class */ (function () {
    function SandboxChannel() {
        this.listeners = [];
        this.messages = [];
        this.closed = false;
    }
    SandboxChannel.prototype.publish = function (message) {
        this.messages.push(message);
        return Promise.resolve();
    };
    SandboxChannel.prototype.notify = function (message) {
        this.listeners.forEach(function (dispatch) { return dispatch(message); });
    };
    SandboxChannel.prototype.subscribe = function (listener) {
        if (!this.listeners.includes(listener)) {
            this.listeners.push(listener);
        }
    };
    SandboxChannel.prototype.unsubscribe = function (listener) {
        var index = this.listeners.indexOf(listener);
        if (index >= 0) {
            this.listeners.splice(index, 1);
        }
    };
    SandboxChannel.prototype.close = function () {
        this.closed = true;
        return Promise.resolve();
    };
    SandboxChannel.prototype.isClosed = function () {
        return this.closed;
    };
    return SandboxChannel;
}());
exports.SandboxChannel = SandboxChannel;
