// @croct/sdk/sourceLocation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLocation = exports.getPosition = exports.getLength = void 0;
var tslib_1 = require("tslib");
function getLength(input) {
    return tslib_1.__spreadArray([], tslib_1.__read(input), false).length;
}
exports.getLength = getLength;
function getPosition(input, index) {
    return getLocation(input, index, index).start;
}
exports.getPosition = getPosition;
function getLocation(input, startIndex, endIndex) {
    if (startIndex < 0) {
        throw Error('The start index cannot be negative.');
    }
    if (endIndex < startIndex) {
        throw new Error('The end index must greater than or equal to the start index.');
    }
    var start;
    var end;
    var chars = tslib_1.__spreadArray([], tslib_1.__read(input), false);
    var line = 1;
    var column = 0;
    for (var offset = 0; offset < chars.length; offset++) {
        var char = chars[offset];
        if (offset === startIndex) {
            start = {
                index: offset,
                line: line,
                column: column,
            };
        }
        if (offset === endIndex) {
            end = {
                index: offset,
                line: line,
                column: column,
            };
            break;
        }
        if (char === '\n') {
            line += 1;
            column = 0;
        }
        else {
            column += 1;
        }
    }
    if (start === undefined) {
        start = {
            index: chars.length,
            line: line,
            column: column,
        };
    }
    if (end === undefined) {
        end = {
            index: chars.length,
            line: line,
            column: column,
        };
    }
    return {
        start: start,
        end: end,
    };
}
exports.getLocation = getLocation;
