// @croct/sdk/trackingEvents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isCartPartialEvent = exports.isIdentifiedUserEvent = exports.eventTypes = exports.miscEventTypes = exports.userEventTypes = exports.identifiedUserEventTypes = exports.ecommerceEventTypes = exports.cartEventTypes = exports.tabEventTypes = exports.pageEventTypes = void 0;
var tslib_1 = require("tslib");
/*
 * Events
 */
exports.pageEventTypes = [
    'pageLoaded',
    'pageOpened',
];
exports.tabEventTypes = [
    'tabOpened',
    'tabUrlChanged',
    'tabVisibilityChanged',
];
exports.cartEventTypes = [
    'cartModified',
    'cartViewed',
    'checkoutStarted',
];
exports.ecommerceEventTypes = tslib_1.__spreadArray(tslib_1.__spreadArray([], tslib_1.__read(exports.cartEventTypes), false), [
    'orderPlaced',
    'productViewed',
], false);
exports.identifiedUserEventTypes = [
    'userSignedIn',
    'userSignedOut',
    'userSignedUp',
];
exports.userEventTypes = tslib_1.__spreadArray(tslib_1.__spreadArray([], tslib_1.__read(exports.identifiedUserEventTypes), false), [
    'userProfileChanged',
], false);
exports.miscEventTypes = [
    'nothingChanged',
    'sessionAttributesChanged',
    'goalCompleted',
    'interestShown',
    'postViewed',
    'eventOccurred',
    'linkOpened',
];
exports.eventTypes = tslib_1.__spreadArray(tslib_1.__spreadArray(tslib_1.__spreadArray(tslib_1.__spreadArray([], tslib_1.__read(exports.pageEventTypes), false), tslib_1.__read(exports.ecommerceEventTypes), false), tslib_1.__read(exports.userEventTypes), false), tslib_1.__read(exports.miscEventTypes), false);
/*
 * Type guards
 */
function isIdentifiedUserEvent(event) {
    return exports.identifiedUserEventTypes.includes(event.type);
}
exports.isIdentifiedUserEvent = isIdentifiedUserEvent;
function isCartPartialEvent(event) {
    return exports.cartEventTypes.includes(event.type);
}
exports.isCartPartialEvent = isCartPartialEvent;
